src/data_lakehouse_ingest/orchestrator/schema_utils.py

```diff
@@ -0,0 +1,374 @@
+"""
+Schema utilities for the Data Lakehouse Ingest framework.
+
+Provides helpers to resolve table schemas, parse SQL-style schema definitions,
+and align DataFrame columns using governed Spark DataTypes, including support
+for DECIMAL(p,s) and ARRAY<T> types.
+"""
+from minio import Minio
+from pyspark.sql import SparkSession, DataFrame
+from pyspark.sql.functions import col, from_json
+from pyspark.sql.types import (
+    StringType,
+    IntegerType,
+    LongType,
+    DoubleType,
+    FloatType,
+    BooleanType,
+    DateType,
+    TimestampType,
+    DecimalType,
+    DataType,
+    ArrayType,
+)
+
+import logging
+from enum import Enum
+import re
+
+class SchemaSource(Enum):
+    """Enum describing the origin of a resolved schema."""
+    SCHEMA_SQL = "schema_sql"
+    INFERRED = "inferred"
+
+
+def resolve_schema(
+    spark: SparkSession,
+    table: dict[str, object],
+    logger: logging.Logger,
+    minio_client: Minio | None = None,
+) -> tuple[str | None, SchemaSource]:
+    """
+    Resolve the schema definition for a given table.
+
+    Current behavior (LinkML not yet supported):
+        - If a LinkML schema path is provided, the function logs an error and
+          raises NotImplementedError. There is **no** automatic fallback to SQL
+          when LinkML is present.
+        - If `schema_sql` is provided (and no LinkML schema), it is returned and
+          marked as SchemaSource.SCHEMA_SQL.
+        - If neither is provided, the schema is treated as inferred and marked
+          as SchemaSource.INFERRED.
+
+    Args:
+        spark (SparkSession): Active Spark session (unused until LinkML is implemented).
+        table (dict): Full table definition from the ingestion config.
+            This dict may include many fields (e.g., name, bronze_path, enabled,
+            schema_sql, linkml_schema, partition_by, drop_extra_columns, etc.).
+            Only the schema-related fields are used in this function.
+        logger (logging.Logger): Logger for reporting resolution decisions.
+        minio_client (Minio | None): Placeholder for future MinIO-based schema retrieval.
+            In the future, callers may want to supply MinIO paths to SQL DDL files,
+            JSON Schemas, or other schema formats stored in MinIO.
+
+    Returns:
+        tuple[str | None, SchemaSource]:
+            - schema_sql (str | None): The resolved SQL-style schema string,
+              or None if the schema is inferred.
```

Comment:
I assume this will change later when the schema string parsing is done earlier in the code path